WIP: feat: Use ServiceAccount to generate token

akrejcir opened the pull request:
**What this PR does / why we need it**:
This PR adds an alternate implementation of the proxy.
- Tokens are generated by kubernetes
- Tokens can be used with `/vnc` subreosurce on Kubevirt `VirtualMachineInstance`.

**Release note**:
```release-note
Added another proxy implementation, that uses ServiceAccounts to generate tokens.
```

ksimon1 commented on pkg/console/config/config.go:
```diff
@@ -1,169 +1,136 @@
-		certsPath: certPath,
-		keyPath:   keyPath,
-		certError: fmt.Errorf("certificate not loaded"),
+		tlsProfileError:  fmt.Errorf("tls profile not loaded"),
```
can you extract the strings as constants?

akrejcir replied:
The string is used only in this place and the unit tests. I don't think a constant is needed.

codingben commented on pkg/console/service/service.go:
```diff
@@ -32,6 +32,21 @@ const (
+		kubevirtClient:  kubevirtClient,
+		metadataClient:  metadataClient,
+		websocketDialer: dialer.New(),
+
```
Why there is an empty line?

codingben commented on pkg/console/service/kubevirt-proxy.go:
```diff
@@ -1,29 +1,253 @@
+		return
+	}
+
+	namespace := request.PathParameter("namespace")
```
Just out of curiosity, is there or will be there some parameters validation? e.g. namespace is empty, and we don't want to pass it to other functions.

akrejcir replied:
I've added validation.

codingben commented on pkg/console/service/kubevirt-proxy.go:
```diff
@@ -1,29 +1,253 @@
+	}
+
+	serviceAccountName := vm.Name + "-vnc-access"
+	err = k.createResources(request.Request.Context(), serviceAccountName, vm)
```
You're passing here `serviceAccountName` to `createResources`, but that `serviceAccountName` is also used as a name for other resources, not only ServiceAccount.

akrejcir replied:
Renamed.

codingben commented on pkg/console/service/kubevirt-proxy.go:
```diff
@@ -1,29 +1,253 @@
+	}
+
+	namespace := request.PathParameter("namespace")
+	name := request.PathParameter("name")
```
I'd change it to `vmName`.

codingben commented on pkg/console/service/kubevirt-proxy.go:
```diff
@@ -1,29 +1,253 @@
+		AppKubernetesNameLabel:      "vm-console-proxy",
+		AppKubernetesPartOfLabel:    "vm-console-proxy",
+		AppKubernetesManagedByLabel: "vm-console-proxy",
+		AppKubernetesComponentLabel: "vm-console-proxy",
```
Can we use constant string that will be stored in `constants.go`?

akrejcir replied:
I've extracted it to a constant, but I didn't to move it to `constants.go`. That file contains constants that are used in multiple other files.

codingben commented on pkg/console/service/kubevirt-proxy_test.go:
```diff
@@ -1,9 +1,221 @@
+		})
+
+		It("should create resources", func() {
+			const resourceName = testName + "-vnc-access"
```
Maybe we can generate `testName + "-vnc-access"` from a function?
```
func generateResourceName(name string) {
return name  + "-vnc-access"
}
```

akrejcir replied:
No, I don't want to call private function from unit tests, and exporting that function just for unit test is unnecessary.

codingben commented on pkg/console/service/kubevirt-proxy_test.go:
```diff
@@ -1,9 +1,221 @@
+		response.SetRequestAccepts(restful.MIME_JSON)
+	})
+
+	Context("TokenHandler", func() {
```
Is there a test to check what happens if we get empty/nil parameters, inside the handler?

akrejcir replied:
DOne

codingben reviewed (commented):
I've added a few comments.

codingben commented on tests/direct-proxy_test.go:
```diff
@@ -30,114 +30,30 @@ import (
+		tokenRequest, err := ApiClient.CoreV1().ServiceAccounts(testNamespace).CreateToken(context.TODO(), serviceAccountName, tokenRequest, metav1.CreateOptions{})
 		Expect(err).ToNot(HaveOccurred())
 
 		saToken = tokenRequest.Status.Token
```
Personally I'd prefer to name it `serviceAccountToken`, and in all related places instead of `sa`. Or just keep it as `token`.

akrejcir replied:
That's not in scope of this PR. We can post another one.

ksimon1 commented on pkg/console/config/config.go:
```diff
@@ -1,169 +1,136 @@
-	case strings.HasPrefix(w.keyPath, event.Name):
-		w.reloadCertificate()
-	}
+	c.reloadTlsProfileLocked()
```
why does these functions have "Locked" in their names?

akrejcir replied:
Because they are called in a critical section when `c.lock` is locked.
Now that I think about it, the whole function does not need to be called with locked lock. I will optimize it in a future PR.

jcanocan commented on pkg/console/config/config.go:
```diff
@@ -1,169 +1,136 @@
 	"kubevirt.io/client-go/log"
 
 	"github.com/kubevirt/vm-console-proxy/api/v1alpha1"
+	"github.com/kubevirt/vm-console-proxy/pkg/filewatch"
```
Did you follow any specific rule to separate the imports? E.g. First block for native go libs, k8s imports, third OpenShift, etc.? Same question for the rest of the files.

akrejcir replied:
Yes:
- Standard library
- Imports from modules in `go.mod`
- Local imports from this project

But my IDE does not enforce it, so sometimes imports are not separated to these blocks.

jcanocan commented on pkg/console/config/config.go:
```diff
@@ -1,169 +1,136 @@
-	Run(done <-chan struct{}) error
+const (
+	SaTokensFilename = "use-service-account-tokens"
+)
```
nit: I would drop the parenthesis since it's a single `const`.

akrejcir replied:
done

jcanocan commented on pkg/console/service/constants.go:
```diff
@@ -0,0 +1,7 @@
+
+const (
+	defaultTokenDuration = 10 * time.Minute
+)
```
nit: remove parenthesis

jcanocan reviewed (commented):
Added a few comments.

akrejcir commented on pkg/console/config/config_test.go:
```diff
@@ -0,0 +1,341 @@
+		Expect(err).ToNot(HaveOccurred())
+
+		Expect(config.CipherSuites).To(ConsistOf(
+			tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
```
I'm not sure about other platforms.
I've added the cipher suites here from openshift API code: https://github.com/kubevirt/vm-console-proxy/blob/039ba0ba60d3fd9ca49d8f1b763d7b4b1e0ab81c/vendor/github.com/openshift/api/config/v1/types_tlssecurityprofile.go#L238-L253

Some of the suites are always enabled, so they are not part of the `CipherSuites` list.

codingben commented on pkg/console/config/config.go:
```diff
@@ -1,169 +1,136 @@
+	return &configWatch{
+		configDir:       configDir,
+		tlsProfileName:  tlsProfileName,
+		useSaTokensName: SaTokensFilename,
```
I think `useSaTokensName` isn't clear what it means, is it contains the file name? If so, can you please name it as `saTokensFilename`?

akrejcir replied:
done

codingben commented on pkg/console/config/config.go:
```diff
@@ -1,169 +1,136 @@
-		keyPath:   keyPath,
-		certError: fmt.Errorf("certificate not loaded"),
+		tlsProfileError:  fmt.Errorf("tls profile not loaded"),
+		useSaTokensError: fmt.Errorf("use-service-account-tokens not loaded"),
```
Can we use here the variable `SaTokensFilename` instead of `use-service-account-tokens`? (also in all other similar places).

akrejcir replied:
done

codingben commented on pkg/console/config/config.go:
```diff
@@ -177,23 +144,51 @@ func (w *watch) reloadTlsProfile() {
+	useSaTokensData, err := os.ReadFile(filepath.Join(c.configDir, c.useSaTokensName))
+	if errors.Is(err, os.ErrNotExist) {
+		c.useSaTokens = false
+		return
```
Since `return` stops the execution, maybe add a log before that the file doesn't exist?

akrejcir replied:
done

akrejcir commented:
The latest revision contains a big change. 
I've removed the old implementation and replaced it with the new one.

/cc @ksimon1 @lyarwood @0xFelix

codingben reviewed (commented):
> The latest revision contains a big change. I've removed the old implementation and replaced it with the new one.
> 
> /cc @ksimon1 @lyarwood @0xFelix

So basically what `vm-console-proxy` provides now is an ability to generate and get a token using `/token` endpoint, so that token can be used to access `VMI/vnc`? Can you please update `docs/api.md` in this PR?

akrejcir commented:
> So basically what `vm-console-proxy` provides now is only an ability to generate and get a token using `/token` endpoint, so that token can be used to access `VMI/vnc`?

Yes. Now it's not a proxy anymore, but it is difficult to change the name of a project. We can consider it for a future change.

> Can you please update `docs/api.md` in this PR?

Updated.

codingben commented on README.md:
```diff
@@ -1,8 +1,8 @@
+The implementation of this project has changed, and now it's not a proxy anymore.
 
-It can generate time limited tokens that are then used to access VNC.
+It provides an endpoint to generate time limited tokens that are then used to access VNC.
```
I'd write it as:
```
# VM Console Proxy

Provides an endpoint to generate time limited tokens that are then used to access VNC.

**Note:*** This project no longer provides VNC proxy functionality, so the project name should be changed in the future.
```

akrejcir replied:
I wrote it differently. If you want, I can change it.

jcanocan commented on tests/tests_suite_test.go:
```diff
@@ -48,22 +64,142 @@ var _ = BeforeSuite(func() {
+		Transport: transport,
+	}
+
+	originalConfigMap, err = ApiClient.CoreV1().ConfigMaps(DeploymentNamespace).Get(context.TODO(), configMapName, metav1.GetOptions{})
```
In the RBAC file, the operator does not have permissions to perform actions on ConfigMaps, how could this work?

akrejcir replied:
Because the functional test code is not executed by the operator. The test binary is run by the CI (or local developer) that has full access to the cluster.

jcanocan replied:
Cool! Thanks for the explanation!

jcanocan commented on tests/tests_suite_test.go:
```diff
@@ -48,22 +64,142 @@ var _ = BeforeSuite(func() {
 			Name: testNamespace,
 		},
 	}
 	_, err = ApiClient.CoreV1().Namespaces().Create(context.TODO(), namespaceObj, metav1.CreateOptions{})
```
Same as configMaps.

jcanocan commented on tests/tests_suite_test.go:
```diff
@@ -48,22 +64,142 @@ var _ = BeforeSuite(func() {
 	_, err = ApiClient.CoreV1().Namespaces().Create(context.TODO(), namespaceObj, metav1.CreateOptions{})
 	Expect(err).ToNot(HaveOccurred())
+	DeferCleanup(func() {
+		err := ApiClient.CoreV1().Namespaces().Delete(context.TODO(), namespaceObj.Name, metav1.DeleteOptions{})
```
Same as above.

akrejcir replied:
Same.

jcanocan reviewed (commented):
Good job here! Added a couple of comments.

lyarwood commented:
> > So basically what `vm-console-proxy` provides now is only an ability to generate and get a token using `/token` endpoint, so that token can be used to access `VMI/vnc`?
> 
> Yes. Now it's not a proxy anymore, but it is difficult to change the name of a project. We can consider it for a future change.

Apologies but I've forgotten why this couldn't be an endpoint in core `KubeVirt` at this point if it's just providing a scoped token for use with the `VMI/VNC` endpoint?

akrejcir commented:
> Apologies but I've forgotten why this couldn't be an endpoint in core `KubeVirt` at this point if it's just providing a scoped token for use with the `VMI/VNC` endpoint?

Because it would require `virt-api` to have permissions to create `ServiceAccounts`, `Roles` and `RoleBindings`, and it could be a security risk. There was design proposal PR: https://github.com/kubevirt/community/pull/193

0xFelix commented on pkg/console/service/service.go:
```diff
@@ -0,0 +1,282 @@
+	})
+}
+
+func (s *service) VncHandler(_ *restful.Request, response *restful.Response) {
```
Do we need really this? Wouldn't implicitly returning 404 be enough?

ksimon1 replied:
In case of upgrade, it is a good idea to indicate to user, that the endpoint was removed

0xFelix replied:
@akrejcir WDYT?

0xFelix commented on pkg/console/service/service.go:
```diff
@@ -0,0 +1,282 @@
+		return fmt.Errorf("failed to create service account: %w", err)
+	}
+
+	role, err := createOrUpdate[*rbac.Role](
```
Maybe instead of creating this role for each token we could make it a ClusterRole that is deployed with vm-console-proxy and drop the privilege to create new roles from vm-console-proxy?

akrejcir replied:
That is exactly what we don't want. A token should only allow access to a single VM.

0xFelix replied:
ACK

0xFelix commented on pkg/console/service/service_test.go:
```diff
@@ -0,0 +1,236 @@
+			testService.TokenHandler(request, response)
+
+			Expect(recorder.Code).To(Equal(http.StatusUnauthorized))
+			Expect(recorder.Body.String()).To(Equal("authenticating token cannot be empty"))
```
IMO it would be better to just return 401 without any further reasons. Do not provide more information than necessary.

akrejcir commented on pkg/console/service/service.go:
```diff
@@ -0,0 +1,282 @@
+	}
+
+	duration := 10 * time.Minute
+	durationParam := request.QueryParameter("duration")
```
Why should we impose an upper bound?

0xFelix replied:
I perceived the generated tokens to be temporary and in such a security relevant context I would expect some kind of reasonable expiration time.

akrejcir replied:
I'm not sure what expiration time is reasonable. I will add a `ConfigMap` option, so users can set it as they want.

akrejcir replied:
I will do that in another PR.

akrejcir commented on pkg/console/service/rbac.go:
```diff
@@ -0,0 +1,89 @@
+	kubevirtv1 "kubevirt.io/api/core/v1"
+)
+
+func ValidateVncRbac(request *restful.Request, cli kubernetes.Interface) error {
```
Good point. I will look into how to implement the extension and see if the change will be worth it.

akrejcir replied:
After working on it for some time, I see that the change will be larger than I expected. I will do it in a future PR.

0xFelix commented on pkg/console/service/service.go:
```diff
@@ -0,0 +1,353 @@
+
+	authToken := getAuthToken(request)
+	if authToken == "" {
+		_ = response.WriteError(http.StatusUnauthorized, fmt.Errorf("authenticating token cannot be empty"))
```
IMO we should not give more info on 401. We could potentially log internally, but over the API I would not return more information to a potential attacker than necessary.

0xFelix commented on pkg/console/service/service.go:
```diff
@@ -0,0 +1,353 @@
+
+	err = s.checkVncRbac(request.Request.Context(), authToken, params.name, params.namespace)
+	if err != nil {
+		_ = response.WriteError(http.StatusUnauthorized, err)
```
Same.

0xFelix commented on pkg/console/service/service.go:
```diff
@@ -0,0 +1,353 @@
+	Get(ctx context.Context, name string, opts metav1.GetOptions) (PT, error)
+}
+
+func createOrUpdate[PT interface {
```
Could this and L332 be part of a library? I see those kind of functions duplicated a lot across our projects.

akrejcir replied:
The `retryOnConflict()` function is nearly a copy of this one from `k8s.io/client-go`: https://github.com/kubernetes/client-go/blob/64a35f6a46ec8a791d437495fd91c87bcd01a5b5/util/retry/util.go#L48-L66

Where are they duplicated?

0xFelix reviewed (commented):
I'm okay with this in general, but please address the comments in follow ups.

/approve

codingben reviewed (approved):
/lgtm
